src/librustc_mir/transform/validate.rs

```diff
@@ -49,6 +53,51 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
         );
     }
 
+    fn check_ty_callable(&self, location: Location, ty: Ty<'tcx>) {
+        if let ty::FnPtr(..) | ty::FnDef(..) = ty.kind {
+            // We have a `FnPtr` or `FnDef` which is trivially safe to call.
+        } else if let Some(fn_once_trait) = self.tcx.lang_items().fn_once_trait() {
+            // FIXME(doctorn): call shims shouldn't reference unnormalized types, so
+            // this case should be removed.
+            // We haven't got a `FnPtr` or `FnDef` but we are still safe to call it if it
+            // implements `FnOnce` (as `Fn: FnMut` and `FnMut: FnOnce`).
+            let item_def_id = self
+                .tcx
+                .associated_items(fn_once_trait)
+                .in_definition_order()
+                .next()
+                .unwrap()
+                .def_id;
+            self.tcx.infer_ctxt().enter(|infcx| {
+                let trait_ref = ty::TraitRef {
+                    def_id: fn_once_trait,
+                    substs: self.tcx.mk_substs_trait(
+                        ty,
+                        infcx.fresh_substs_for_item(self.body.span, item_def_id),
+                    ),
+                };
+                let predicate = ty::PredicateKind::Trait(
+                    ty::Binder::bind(ty::TraitPredicate { trait_ref }),
+                    Constness::NotConst,
+                )
+                .to_predicate(self.tcx);
+                let obligation = traits::Obligation::new(
+                    traits::ObligationCause::dummy(),
+                    self.param_env,
+                    predicate,
+                );
+                if !infcx.predicate_may_hold(&obligation) {
+                    self.fail(
+                        location,
+                        format!("encountered non-callable type {} in `Call` terminator", ty),
+                    );
+                }
+            });
+        } else {
+            self.fail(location, format!("missing `FnOnce` lang-item so cannot verify callability"));
```

Comment:
Why not use `require_lang_item`?

Reply:
I was thinking in terms of it being hypothetically possible to write code that compiles without requiring `FnOnce` in `#[no_core]` and therefore it seemed weird to put in a requirement for it in a validation pass. I guess it’s extremely unlikely that would ever be the case though...

Reply:
But now shims will cause an ICE when `FnOnce` is missing, right? `require_lang_item` instead raises a fatal error, which seems like reasonable behavior when reaching this code.

Reply:
Yeah - that’s true - I can update this tomorrow if you think it’s worth it but this definitely needs to be removed anyway.

Reply:
Yes, that would be good.